bindings/c/test/apitester/TesterBlobGranuleErrorsWorkload.cpp

```diff
@@ -0,0 +1,153 @@
+
+namespace FdbApiTester {
+
+class BlobGranuleErrorsWorkload : public ApiWorkload {
+public:
+	BlobGranuleErrorsWorkload(const WorkloadConfig& config) : ApiWorkload(config) {}
+
+private:
+	enum OpType {
+		OP_READ_NO_MATERIALIZE,
+		OP_READ_FILE_LOAD_ERROR,
+		OP_READ_TOO_OLD,
+		OP_GRANULE_READ_CALLBACK_BLOCKED,
+		OP_CANCEL_RANGES,
+		OP_LAST = OP_CANCEL_RANGES
+	};
+
+	void doErrorOp(TTaskFct cont,
+	               std::string basePathAddition,
+	               bool doMaterialize,
+	               int64_t readVersion,
+	               fdb_error_t expectedError) {
+		std::string begin = randomKeyName();
+		std::string end = randomKeyName();
+		if (begin > end) {
+			std::swap(begin, end);
+		}
+
+		execTransaction(
+		    [begin, end, basePathAddition, doMaterialize, readVersion, expectedError](auto ctx) {
+			    ctx->tx()->setOption(FDB_TR_OPTION_READ_YOUR_WRITES_DISABLE);
+			    KeyValuesResult res = ctx->tx()->readBlobGranules(
+			        begin, end, ctx->getBGBasePath() + basePathAddition, doMaterialize, readVersion);
+			    bool more;
+			    res.getKeyValues(&more);
+
+			    ASSERT(res.getError() != error_code_success);
+			    if (res.getError() != expectedError) {
+				    ctx->onError(res.getError());
+			    } else {
+				    ctx->done();
+			    }
+		    },
+		    [this, cont]() { schedule(cont); });
+	}
+
+	void randomOpReadNoMaterialize(TTaskFct cont) {
+		// ensure setting noMaterialize flag produces blob_granule_not_materialized
+		doErrorOp(cont, "", false, -2, error_code_blob_granule_not_materialized);
+	}
+
+	void randomOpReadFileLoadError(TTaskFct cont) {
+		// point to a file path that doesn't exist by adding an extra suffix
+		doErrorOp(cont, "extrapath/", true, -2, error_code_blob_granule_file_load_error);
+	}
+
+	void randomOpReadTooOld(TTaskFct cont) {
+		// read at a version (1) that predates granule data
+		doErrorOp(cont, "extrapath/", true, 1, error_code_blob_granule_transaction_too_old);
+	}
+
+	// Do a get, and in its callback do a readBlobGranules. Because readBlobGranules is synchronous and does a
+	// blockOnMainThread, it should get the error_code_blocked_from_network_thread error to avoid deadlocks.
+	void randomReadCallbackBlocked(TTaskFct cont) {
+		std::string begin = randomKeyName();
+		std::string end = randomKeyName();
+		if (begin > end) {
+			std::swap(begin, end);
+		}
+		execTransaction(
+		    [begin, end](auto ctx) {
+			    ctx->tx()->setOption(FDB_TR_OPTION_READ_YOUR_WRITES_DISABLE);
+			    // this key doesn't matter, just that readBlobGranules is in the callback
+			    auto f = ctx->tx()->get(begin, false);
+			    ctx->continueAfter(
+			        f,
+			        [ctx, f, begin, end]() {
+				        KeyValuesResult res = ctx->tx()->readBlobGranules(begin, end, ctx->getBGBasePath());
+				        bool more;
+				        res.getKeyValues(&more);
+				        ASSERT(res.getError() != error_code_success);
+				        if (res.getError() != error_code_blocked_from_network_thread) {
```

Comment:
The continuation given in continueAfter is not executed on the main thread, but rather scheduled to be executed on the client thread pool. Do you really get blocked_from_network_thread error here?

Reply:
That makes sense.
The purpose of this test is that, if a client does a blocking operation on a network thread that would deadlock it, it is desirable for them to get an error instead of deadlocking the thread.
We already have this implemented in ThreadHelper.actor.h for callbacks on the same network thread.
But @ajbeamon suggested we add a separate check for deadlocks for readBlobGranules, since it returns a synchronous result instead of an async future, and does a blocking call to the network thread to do so. In particular, this deadlock guard was designed to prevent against the callback executing against a different network thread.
My goal was for the randomReadCallbackBlocked test to reproduce the deadlock and blocking error, so that we could be sure the deadlock guard I added fixes the problem.